src/main/java/org/apache/sysds/hops/cost/FederatedCostEstimator.java

```diff
@@ -196,6 +194,90 @@ private FederatedCost costEstimate(Hop root){
 		}
 	}
 
+	/**
+	 * Return cost estimate in bytes of Hop DAG starting from given root HopRel.
+	 * @param root HopRel of Hop DAG for which cost is estimated
+	 * @param hopRelMemo memo table of HopRels for calculating input costs
+	 * @return cost estimation of Hop DAG starting from given root HopRel
+	 */
+	public FederatedCost costEstimate(HopRel root, Map<Long, List<HopRel>> hopRelMemo){
+		// Check if root is in memo table.
+		if ( hopRelMemo.containsKey(root.hopRef.getHopID())
+			&& hopRelMemo.get(root.hopRef.getHopID()).stream().anyMatch(h -> h.fedOut == root.fedOut) ){
+			return root.getCostObject();
+		}
+		else {
+			// If no input has FOUT, the root will be processed by the coordinator
+			boolean hasFederatedInput = root.inputDependency.stream().anyMatch(in -> in.hopRef.hasFederatedOutput());
+			//the input cost is included the first time the input hop is used
+			//for additional usage, the additional cost is zero (disregarding potential read cost)
+			double inputCosts = root.inputDependency.stream()
+				.mapToDouble( in -> {
+					double inCost = in.existingCostPointer(root.hopRef.getHopID()) ?
+						0 : costEstimate(in, hopRelMemo).getTotal();
+					in.addCostPointer(root.hopRef.getHopID());
+					return inCost;
+				} )
+				.sum();
```

Comment:
seems overkill with a stream sum here, i would suggest making method, where you use a for loop instead.

Reply:
if you don't do that i would suggest that you define a method that you use int the map to double instead of the lambda function defined here.